File: go/common/api_credentials/credentials.go
package api_credentials

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

const MESH_API_CREDENTIALS_ENV_VAR = "MESH_API_CREDENTIALS"

// Credentials represents API credentials loaded from file
type Credentials struct {
	APIKey string `json:"api_key"`
	Group  string `json:"group"`
}

// LoadCredentialsFromFile loads API credentials from a JSON file
func LoadCredentialsFromFile(path string) (*Credentials, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read credentials file: %w", err)
	}

	var creds Credentials
	if err := json.Unmarshal(data, &creds); err != nil {
		return nil, fmt.Errorf("failed to parse credentials file: %w", err)
	}

	if creds.APIKey == "" {
		return nil, fmt.Errorf("api_key is required in credentials file")
	}

	if creds.Group == "" {
		return nil, fmt.Errorf("group is required in credentials file")
	}

	if !strings.HasPrefix(creds.Group, "groups/") {
		return nil, fmt.Errorf("group must be in format groups/{group_id}, got: %s", creds.Group)
	}

	return &creds, nil
}

// CredentialsFromEnvironment loads API credentials from the file path specified in MESH_API_CREDENTIALS environment variable
func CredentialsFromEnvironment() (*Credentials, error) {
	path := os.Getenv(MESH_API_CREDENTIALS_ENV_VAR)
	if path == "" {
		return nil, fmt.Errorf("MESH_API_CREDENTIALS environment variable not set")
	}

	return LoadCredentialsFromFile(path)
}

File: python/src/meshtrade/common/api_credentials/credentials.py
"""
API Credentials loading functionality for Meshtrade SDK.
"""

import json
import os

from meshtrade.iam.api_user.v1.api_credentials_pb2 import APICredentials

MESH_API_CREDENTIALS_ENV_VAR = "MESH_API_CREDENTIALS"


def load_credentials_from_file(path: str) -> APICredentials:
    """Load API credentials from a JSON file.

    Args:
        path: Path to the credentials JSON file

    Returns:
        APICredentials object with loaded credentials

    Raises:
        FileNotFoundError: If the credentials file doesn't exist
        ValueError: If the credentials file is invalid or missing required fields
    """
    try:
        with open(path) as f:
            data = json.load(f)
    except FileNotFoundError as e:
        raise FileNotFoundError(f"Failed to read credentials file: {path}") from e
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse credentials file: {e}") from e

    api_key = data.get("api_key")
    group = data.get("group")

    if not api_key:
        raise ValueError("api_key is required in credentials file")

    if not group:
        raise ValueError("group is required in credentials file")

    if not group.startswith("groups/"):
        raise ValueError(f"group must be in format groups/{{group_id}}, got: {group}")

    creds = APICredentials()
    creds.api_key = api_key
    creds.group = group
    
    return creds


def credentials_from_environment() -> APICredentials | None:
    """Load API credentials from the file path specified in MESH_API_CREDENTIALS environment variable.

    Returns:
        APICredentials object if environment variable is set and file exists, None otherwise

    Raises:
        ValueError: If the credentials file is invalid
    """
    path = os.getenv(MESH_API_CREDENTIALS_ENV_VAR)
    if not path:
        return None

    return load_credentials_from_file(path)
